Multitask Whisper PR

siddhu001 opened the pull request:
## What?

Add code for Multitasking SLU tasks for Whisper formulation

## Why?

Add code for Multitasking SLU tasks for Whisper formulation

sw005320 commented:
@siddhu001, can you try to follow the recipe naming convention as much as possible?
We usually do not use the upper case. 
Also, `asr1_classification_multitask` is too long.

ftshijt commented on egs2/ASVSpoof/asr1/local/data.sh:
```diff
@@ -0,0 +1,62 @@
+. ./db.sh
+. ./path.sh
+. ./cmd.sh
+ASVSpoof_CMD=/scratch/bbjs/shared/corpora/ASVSpoof
```
Do we want this hardcoded?

ftshijt commented on egs2/ASVSpoof/asr1/local/data.sh:
```diff
@@ -0,0 +1,62 @@
+if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
+    if [ ! -e "${ASVSpoof_CMD}/README.LA.txt" ]; then
+	    echo "stage 1: Download data to ${ASVSpoof_CMD}"
+        # git clone https://github.com/kolesov93/lt_speech_commands.git ${LT_SPEECH_CMD}
```
Let's uncomment it

ftshijt commented on egs2/ASVSpoof/asr1/local/data.sh:
```diff
@@ -0,0 +1,62 @@
+    python3 local/data_prep.py ${ASVSpoof_CMD}
+    for x in test valid train; do
+        for f in text wav.scp utt2spk; do
+            echo ${x}, ${f}
+            echo "?"
+            head data/${x}/${f}
```
Sorry, I do not fully understand the logic here. Could you explain a bit?

ftshijt commented on egs2/ASVSpoof/asr1/local/data_prep.py:
```diff
@@ -0,0 +1,49 @@
+            cls=line.split(" ")[4].replace("\n", "")
+            spk=line.split(" ")[0]
+            utt_id=spk+"_"+line.split(" ")[1]
+            # utt_id, cls = line.strip().split("\t")
+            # spk=utt_id.split("/")[-1]
+            # utt_id = "/".join(utt_id.split("/")[-2:])
+            data_dir =  Path(split_path, f'{line.split(" ")[1]}.flac')
+            # utt_id = spk+"_"+ cls + "_" + utt_id
```
Please remove the commented lines

ftshijt commented on egs2/ASVSpoof/asr1/run.sh:
```diff
@@ -0,0 +1,33 @@
+    --lang en \
+    --ngpu 1 \
+    --use_lm false \
+    --nbpe 5000 \
```
nbpe won't be activated with token_type to word

ftshijt commented on egs2/Mustard/asr1/local/data.sh:
```diff
@@ -0,0 +1,62 @@
+. ./db.sh
+. ./path.sh
+. ./cmd.sh
+ASVSpoof_CMD=/scratch/bbjs/shared/corpora/MUStARD
```
please remove the hard-coded path

ftshijt commented on egs2/Mustard/asr1/local/data.sh:
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env bash
```
It seems to be identical to ASVSPoof recipe (while only the dataset path is changed)?

ftshijt commented on egs2/Mustard/asr1/local/data_prep.py:
```diff
@@ -0,0 +1,67 @@
+    print("Usage: python data_prep.py [ASVSpoof_root]")
+    sys.exit(1)
+mustard_root = sys.argv[1]
+split_file = os.path.join("/scratch/bbjs/shared/corpora/SpeechPrompt-v2/GSLM/preprocess/SD_mustard/", "split_indices.p")
```
hard-coded path

ftshijt commented on egs2/Mustard/asr1/local/data_prep.py:
```diff
@@ -0,0 +1,67 @@
+train_fnames, val_fnames = random_split(train_val_fnames, [train_len, val_len])
+train_fnames = list(train_fnames)
+val_fnames = list(val_fnames)
+# print(test_fnames[:10])
+# print(train_fnames[:10])
+# print(val_fnames[:10])
+# exit()
```
```suggestion
```

ftshijt commented on egs2/Mustard/asr1/local/data_prep.py:
```diff
@@ -0,0 +1,67 @@
+                cls='class:not_sarcasm'
+            spk=data[line]['show']+"_"+data[line]['speaker']+"_speaker"
+            utt_id=spk+"_"+line
+            # utt_id, cls = line.strip().split("\t")
+            # spk=utt_id.split("/")[-1]
+            # utt_id = "/".join(utt_id.split("/")[-2:])
+            data_dir =  "/scratch/bbjs/shared/corpora/Mustard/utterances_final_wav/"+line+".wav"
+            # utt_id = spk+"_"+ cls + "_" + utt_id
+
```
hard code and useless commented lines

ftshijt commented on egs2/Mustard_plus_plus/asr1/local/data.sh:
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env bash
```
Same issues as ASVSpoof recipe

ftshijt commented on egs2/stop/asr1_classification_multitask/conf/decode_asr_er.yaml:
```diff
@@ -0,0 +1,8 @@
+beam_size: 1
```
Is it necessary to keep so many identical files? Maybe just use symlink

siddhu001 replied:
I do think we need many decoding files because each of them have different task specifiers as prompts.

ftshijt reviewed (changes requested):
Great effort! But most recipes need a bit more cleaning to be able to merge.

sw005320 commented:
@siddhu001, can you restart it?

siddhu001 commented:
@sw005320 I believe we can proceed with merging this PR. I have addressed all the feedback, and the changes pass all the CI checks. Additionally, I've included unit tests to ensure the reliability of the code. Please review and let me know if you have any comments or concerns. I plan to add the results to the README and upload the models to Hugging Face in a follow-up PR.

sw005320 commented on egs2/stop/asr1_multitask/check_vad_results.py:
ditto

sw005320 commented on egs2/stop/asr1_multitask/clean_emotion_pred.py:
ditto

sw005320 commented on egs2/stop/asr1_multitask/conf/decode_asr_ar.yaml:
```diff
@@ -0,0 +1,8 @@
+minlenratio: 0.
+penalty: 0.
+lang_prompt_token: "<|ar|> <|scr|> <|ar_scr|>"
+prompt_token_file: /projects/bbjs/arora1/new_download/espnet/egs2/stop/asr3_less_combined2_more_multilingual_data2/add_tokens-Copy1.txt
```
Why this is an absolute path?
Many config files under this directory are an absolute path.

sw005320 commented on egs2/stop/asr1_multitask/create_better_accentdb.py:
ditto

sw005320 commented on egs2/stop/asr1_multitask/create_better_arabic_scr.py:
ditto

sw005320 commented on egs2/stop/asr1_multitask/create_better_asvspoof.py:
ditto

sw005320 commented on egs2/stop/asr1_multitask/create_better_emotion.py:
ditto

sw005320 commented on egs2/stop/asr1_multitask/create_better_esc50.py:
ditto

sw005320 commented on egs2/voxforge/lid1/convert2audio.py:
ditto

sw005320 reviewed (commented):
- please add the short descriptions about the database to https://github.com/espnet/espnet/blob/master/egs2/README.md (later?)
- there are a lot of similar decoder configs under `egs2/stop/asr1_multitask/conf`. Are they necessary? Can you check and remove unnecessary ones?
- please only put the minimum files under `egs2/stop/asr1_multitask/`
- please add the documentation of how to augment `asr` to accept prompts in details

sw005320 reviewed (commented):
- Can you comp up with another name for "create_better"? This is not concrete.
- `egs2/stop/asr1_multitask/README.md` should be changed according to the change of the files 
- would it be possible to automate this step as much as possible (in the later PR?)? This sounds a bit complicated to follow.

siddhu001 commented:
Hi Shinji,

I appreciate your detailed comments. Here's an update on my progress regarding your suggestions:
1. `Can you comp up with another name for "create_better"? This is not concrete`.: Sure I will change it into create_prompt_*
2. `egs2/stop/asr1_multitask/README.md should be changed according to the change of the files`: I'll make sure to update it to reflect the changes in the filenames.
3. `would it be possible to automate this step as much as possible (in the later PR?)? This sounds a bit complicated to follow.`:  I acknowledge your concern, and I'll work on automating this step in a later PR to make it more user-friendly.
4. `please add the documentation of how to augment asr to accept prompts in details`: I'm actively addressing this comment. I plan to add a README_multitask.md in the asr1 template, following a similar format to https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr2/README.md. This will provide detailed instructions on how to augment ASR1 for prompt-based multi-task learning.
5. `please add the short descriptions about the database to https://github.com/espnet/espnet/blob/master/egs2/README.md`: I'm currently working on this addition as well.

Thank you for your feedback and patience. If you have any further suggestions or questions, please feel free to let me know.

siddhu001 commented:
@sw005320 I think I have incorporated all your suggestions to PR. Let me know in case there are any comments!

sw005320 commented on egs2/stop/asr1_multitask/README.md:
- This document and egs2/TEMPLATE/asr1/README_multitask.md should be merged (or at least cross-referenced).
- Is this the right place to put the multitask recipe? I feel strange if we put them under the `stop` directory. How about naming this task (e.g., uslu17 (17 means the number of tasks)) and putting them under the directory?
  - then, we may not have to have a special name like `asr1_multitask`? Simply `asr1` might be fine.

sw005320 reviewed (commented):
One more comment on the directory structure.
We're almost there.

siddhu001 commented:
Hi @sw005320, 

I can add scoring function from local to egs2/TEMPLATE/asr1/pyscripts/utils in maybe later PR.

Unfortunately, the PR that has just been merged has made it challenging to resolve the merge conflicts for my pull request. This is because it hard codes "whisper task" as "transcribe" in asr.sh, and this value is passed through all whisper tokenizer files. I'll need to make significant changes to the code to add it as an option so that we can run multi-tasking across different tasks in asr.sh. Hopefully, I'll have it working soon.

sw005320 commented:
I'm still not sure that we will have a lot of directories under the main recipe directory (e.g., `dump_*`).
If it is necessary, you can keep them, but if you can move to the other parts (e.g., `dump/{grabo, ...,}`) that is my preference.

siddhu001 commented on egs2/uslu14/asr1/local/prepare_stop_data.py:
Goo catch! Removed this!

siddhu001 commented on egs2/uslu14/asr1/local/score_slurp.sh:
At this time, SLURP is not required because it is not included among the 14 datasets considered for UniverSLU-14. Therefore, it has been removed.

As mentioned in the README at https://github.com/espnet/espnet/blob/94ad0db58545ac9d277c37f9fc9b64c4aad8c09e/egs2/TEMPLATE/asr1/README_multitask.md, some of the scoring functions used to compute the F1 score require manual execution. I plan to automate both the scoring functions and the creation of a dump directory in a follow-up PR.

siddhu001 commented on egs2/uslu14/asr1/local/score_snips.sh:
Again, at this time, SNIPS is not required because it is not included among the 14 datasets considered for UniverSLU-14. Therefore, it has been removed.

siddhu001 commented on espnet2/text/whisper_token_id_converter.py:
```diff
@@ -48,6 +51,14 @@ def __init__(
             self.tokenizer = whisper.tokenizer.get_tokenizer(
                 multilingual=True, language=language, task=task
             )
+            # import pdb;pdb.set_trace()
```
Done!

siddhu001 commented:
> I'm still not sure that we will have a lot of directories under the main recipe directory (e.g., `dump_*`). If it is necessary, you can keep them, but if you can move to the other parts (e.g., `dump/{grabo, ...,}`) that is my preference.

Sounds good. I added them all inside dump directories.

siddhu001 commented:
Hi @sw005320, can we merge this PR now?

sw005320 commented:
OK, thanks!
I merged it.
